src/neo/SmartContract/Native/Tokens/Nep11Token.cs

```diff
@@ -0,0 +1,270 @@
+                        }
+                    },
+                    ReturnType = ContractParameterType.Boolean
+                }
+            };
+
+            Manifest.Abi.Events = events.ToArray();
+        }
+
+        [ContractMethod(0, ContractParameterType.String, CallFlags.None, Name = "name")]
+        protected StackItem NameMethod(ApplicationEngine engine, Array args)
+        {
+            return Name;
+        }
+
+        [ContractMethod(0, ContractParameterType.String, CallFlags.None, Name = "symbol")]
+        protected StackItem SymbolMethod(ApplicationEngine engine, Array args)
+        {
+            return Symbol;
+        }
+
+        [ContractMethod(0, ContractParameterType.Integer, CallFlags.None, Name = "decimals")]
+        protected StackItem DecimalsMethod(ApplicationEngine engine, Array args)
+        {
+            return (uint)Decimals;
+        }
+
+        [ContractMethod(0_01000000, ContractParameterType.Integer, CallFlags.AllowStates)]
+        protected StackItem TotalSupply(ApplicationEngine engine, Array args)
+        {
+            return TotalSupply(engine.Snapshot);
+        }
+
+        public virtual BigInteger TotalSupply(StoreView snapshot)
+        {
+            StorageItem storage = snapshot.Storages.TryGet(CreateStorageKey(Prefix_TotalSupply));
+            if (storage is null) return BigInteger.Zero;
+            return new BigInteger(storage.Value);
+        }
+
+        [ContractMethod(0_01000000, ContractParameterType.Integer, CallFlags.AllowStates, ParameterTypes = new[] { ContractParameterType.Hash160, ContractParameterType.ByteArray }, ParameterNames = new[] { "account", "tokenId" })]
+        protected StackItem BalanceOf(ApplicationEngine engine, Array args)
+        {
+            UInt160 account = new UInt160(args[0].GetSpan());
+            byte[] tokenId = args[1].GetSpan().ToArray();
+            return BalanceOf(engine.Snapshot, account, tokenId);
+        }
+
+        public virtual BigInteger BalanceOf(StoreView snapshot, UInt160 account, byte[] tokenId)
+        {
+            UInt160 innerKey = GetInnerKey(tokenId);
+            StorageItem storage = snapshot.Storages.TryGet(CreateOwnershipKey(account, innerKey));
+            if (storage is null) return BigInteger.Zero;
+            return storage.GetInteroperable<UState>().Balance;
+        }
+
+        [ContractMethod(0_01000000, ContractParameterType.Array, CallFlags.AllowStates, ParameterTypes = new[] { ContractParameterType.Hash160 }, ParameterNames = new[] { "owner" })]
+        public StackItem TokensOf(ApplicationEngine engine, Array args)
+        {
+            UInt160 owner = new UInt160(args[0].GetSpan());
+            return new InteropInterface(TokensOf(engine.Snapshot, owner));
+        }
+
+        public virtual IEnumerator<TState> TokensOf(StoreView snapshot, UInt160 owner)
+        {
+            return snapshot.Storages.Find(CreateStorageKey(Prefix_Ownership, owner).ToArray()).Select(p =>
```

Comment:
I don't know what changes have been applied to the NEOVM limits (such as maximum amount of stack objects or maximum amount of data on the stack) in Neo3, but, if something similar to the limits present in Neo2 is still there, wouldn't it be possible to trigger those limits by transferring a lot of tokens to a specific account, forcing calls to TokensOf to fail?

This is most probably not an issue if this is just being called by clients, but, if some smart contract calls this function, it could trigger unexpected failures.

Reply:
I think the `TokensOf` is not for other contracts, it's for clients. As `TokensOf` is not efficient in contract, the other contracts should use `BalanceOf`.

Reply:
The  iterator operation in the contract is charged. If too many times and the fee exceeds the payment, the execution will be automatically stopped.In addition, there is indeed a stack depth limit in the VM, but we operate on iterators, and when we get a new object, the old object is released.

In addition, it is not wise to obtain such a large amount of data on the chain

Reply:
@doubiliu I see, then, if the consuming contract is only charged for iterator operations, wouldn't it be possible to execute the following attack, which makes it possible to trigger arbitrarily expensive computations for a fixed cost (destroying the symmetry between gas cost and computational expense):
1. Send a lot of tokens to the same owner through different transactions.
2. From another contract, call TokensOf(owner), leave the iterator on the stack and call TokensOf(owner) again repeatedly
3. For each call to TokensOf, a map over the array of tokens is executed, which will consume an arbitrarily high amount of computing power because the length of the array is unbounded, yet the consuming contract will only be charged a fixed amount. Furthermore, all the arrays of tokens that were allocated due to TokensOf(owner) calls need to stay in the memory while the iterators remain in the stack.

Note that this attack also makes it possible to increase a node's memory consumption to an arbitrarily high number (which can lead to a crash) but that attack might not be practical because the arrays only store the ids now.

Reply:
Agree, it will create an attack vector